src/object/srv_obj.c

```diff
@@ -2043,25 +2042,19 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
 		rc = dtx_end(&dth, ioc.ioc_coc, rc);
 
 		if (rc == -DER_INPROGRESS && dth.dth_local_retry) {
-			if (++retry > 3)
+			if (++retry > 10)
 				D_GOTO(out, rc = -DER_TX_BUSY);
 
 			/* XXX: Currently, we commit the distributed transaction
 			 *	sychronously. Normally, it will be very quickly.
-			 *	So let's wait on the server for a while (30 ms),
-			 *	then retry. If related distributed transaction
-			 *	is still not committed after several cycles try,
-			 *	then replies '-DER_TX_BUSY' to the client.
+			 *	So let's yield then retry. If related
+			 *	distributed transaction is still not committed
+			 *	after several cycles, replies '-DER_TX_BUSY' to
+			 *	the client.
 			 */
-			if (sleep_ult == NULL) {
-				sleep_ult = dss_sleep_ult_create();
-				if (sleep_ult == NULL)
-					D_GOTO(out, rc = -DER_TX_BUSY);
-			}
-
 			D_DEBUG(DB_IO, "Hit non-commit DTX when fetch "
 				DF_UOID" (%d)\n", DP_UOID(orw->orw_oid), retry);
-			dss_ult_sleep(sleep_ult, 30000);
+			ABT_thread_yield();
```

Comment:
The yield time is uncontrollable, so if the system is heavy loaded, ten cycles server local retry may take some long time as to increase the possibility of client-side RPC timeout. If you have other chance to refresh the patch, I would suggest to shrink the server local retry times.